cli/internal/publish/images_test.go

```diff
@@ -0,0 +1,220 @@
+		},
+		{
+			name: "basic with title or alt, multiple",
+			contents: `# Title
+![](image.png "Title Here")
+More test
+![alt](image.png)
+`,
+			expect: map[string][]imageReference{"a94a8fe5ccb19ba61c4c0873d391e987982fbbd3:image.png": {{ref: "image.png", startPos: 8, endPos: 36}, {ref: "image.png", startPos: 46, endPos: 63}}},
+		},
+		{
+			name: "ref",
+			contents: `# Title
+![Alt text][image-id]
+
+[image-id]: image.png "Optional Title Here"
+`,
+			expect:    map[string][]imageReference{"a94a8fe5ccb19ba61c4c0873d391e987982fbbd3:image.png": {{ref: "image.png"}}}, // issue: startpos/endpos don't point to file
+			skipCheck: true,
+		},
+		{
+			name: "ref multiple",
+			contents: `# Title
+![Alt text][image-id]
+
+text
+
+![Same text][image-id]
+
+[image-id]: image.png "Optional Title Here"
+`,
+			expect:    map[string][]imageReference{"a94a8fe5ccb19ba61c4c0873d391e987982fbbd3:image.png": {{ref: "image.png"}, {ref: "image.png"}}}, // two references to same image. startpos/endpos issue as above
+			skipCheck: true,
+		},
+		{
+			name: "href",
+			contents: `# Title
+[![Alt text](image.png)](http://example.com/)
+`,
+			expect: map[string][]imageReference{"a94a8fe5ccb19ba61c4c0873d391e987982fbbd3:image.png": {{ref: "image.png", startPos: 8, endPos: 53}}}, // includes full href
+		},
+		{
+			name: "subdir",
+			contents: `# Title
+![Alt text](assets/images/image.png)
+`,
+			expect: map[string][]imageReference{"a94a8fe5ccb19ba61c4c0873d391e987982fbbd3:image.png": {{ref: "assets/images/image.png", startPos: 8, endPos: 44}}},
+		},
+		{
+			name: "basic file://",
+			contents: `# Title
+![](file://${ABS_IMAGE})`,
+			expect: map[string][]imageReference{"a94a8fe5ccb19ba61c4c0873d391e987982fbbd3:my special@image.png": {{ref: "file://${ABS_IMAGE}", startPos: 8, endPos: -12}}}, // 12 is the number of extra characters except the placeholder/filename
+		},
+	}
+
+	tempdir := t.TempDir()
+	data := []byte("test")
+	require.NoError(t, os.WriteFile(filepath.Join(tempdir, "image.png"), data, 0644))
+	require.NoError(t, os.MkdirAll(filepath.Join(tempdir, "assets", "images"), 0755))
+	require.NoError(t, os.WriteFile(filepath.Join(tempdir, "assets", "images", "image.png"), data, 0644))
+
+	special := filepath.Join(tempdir, "my special@image.png")
+	require.NoError(t, os.WriteFile(special, data, 0644))
+	specialAbs, err := filepath.Abs(special)
+	require.NoError(t, err)
+	specialAbsEscaped := strings.NewReplacer(" ", "%20", "@", "%40", string(filepath.Separator), "/").Replace(specialAbs)
+	if filepath.Separator == '\\' {
+		specialAbsEscaped = "/" + specialAbsEscaped // absolute paths in file:// mode start with "/", not C:\ (`file:///C:/`, not `file://C:/`)
+	}
+
+	const mockURL = "https://example.com/file.ext"
+	for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			tc.contents = strings.ReplaceAll(tc.contents, "${ABS_IMAGE}", specialAbsEscaped)
+			out, err := findMarkdownImages(tc.contents, tempdir)
```

Comment:
It feels like it would simplify the validation of test cases if we tested `findMarkdownImages` and `replaceMarkdownImages` in separate tests?

Because if I'm reading this test right, you need to now adjust the startPos and endPos to what it will be *after* replacement, rather than before. I'm also a bit confused by how endPos can be negative, maybe I misunderstand what it's doing 🤔

Reply:
No the positions are before replacement, except for the ABS_IMAGE case, which we don't and can't know the exact path before running the test due to the nature of temp dirs.

Reply:
Why are positions negative sometimes?

Reply:
They aren't really, clarified in https://github.com/cloudquery/cloudquery/pull/16033/commits/aeb4f25e986159dc469a9b205a45110d798a819d